src/gainmap.c

```diff
@@ -49,4 +53,238 @@ avifBool avifGainMapMetadataFractionsToDouble(avifGainMapMetadataDouble * dst, c
+
+    if (weight == 0. && outputTransferCharacteristics == transferCharacteristics &&
+        baseImage->format == toneMappedImage->format && baseImage->isFloat == toneMappedImage->isFloat) {
+        assert(baseImage->rowBytes == toneMappedImage->rowBytes);
+        assert(baseImage->height == toneMappedImage->height);
+        // Copy the base image.
+        memcpy(toneMappedImage->pixels, baseImage->pixels, baseImage->rowBytes * baseImage->height);
+        goto cleanup;
+    }
+
+    if (gainMap->image->width != width || gainMap->image->height != height) {
+        rescaledGainMap = avifImageCreateEmpty();
+        res = avifImageCopy(rescaledGainMap, gainMap->image, AVIF_PLANES_YUV);
+        if (res != AVIF_RESULT_OK) {
+            goto cleanup;
+        }
+        res = avifImageScale(rescaledGainMap, width, height, diag);
+        if (res != AVIF_RESULT_OK) {
+            goto cleanup;
+        }
+    }
+
+    avifRGBImageSetDefaults(&rgbGainMap, rescaledGainMap != NULL ? rescaledGainMap : gainMap->image);
+    res = avifRGBImageAllocatePixels(&rgbGainMap);
+    if (res != AVIF_RESULT_OK) {
+        goto cleanup;
+    }
+    res = avifImageYUVToRGB(rescaledGainMap != NULL ? rescaledGainMap : gainMap->image, &rgbGainMap);
+    if (res != AVIF_RESULT_OK) {
+        goto cleanup;
+    }
+
+    avifRGBColorSpaceInfo gainMapRGBInfo;
+    avifRGBColorSpaceInfo baseRGBInfo;
+    avifRGBColorSpaceInfo toneMappedPixelRGBState;
+    if (!avifGetRGBColorSpaceInfo(&rgbGainMap, &gainMapRGBInfo) || !avifGetRGBColorSpaceInfo(baseImage, &baseRGBInfo) ||
+        !avifGetRGBColorSpaceInfo(toneMappedImage, &toneMappedPixelRGBState)) {
+        avifDiagnosticsPrintf(diag, "Unsupported RGB color space");
+        res = AVIF_RESULT_NOT_IMPLEMENTED;
+        goto cleanup;
+    }
+
+    const avifTransferFunction gammaToLinear = avifTransferCharacteristicsGetGammaToLinearFunction(transferCharacteristics);
+    const avifTransferFunction linearToGamma = avifTransferCharacteristicsGetLinearToGammaFunction(outputTransferCharacteristics);
+
+    float gainMapMinLog[3];
+    float gainMapMaxLog[3];
+    for (int i = 0; i < 3; ++i) {
+        gainMapMinLog[i] = (float)log(metadata.gainMapMin[i]);
+        gainMapMaxLog[i] = (float)log(metadata.gainMapMax[i]);
+    }
+
+    const double * offsetBase = gainMap->metadata.baseRenditionIsHDR ? metadata.offsetHdr : metadata.offsetSdr;
+    const double * offsetOther = gainMap->metadata.baseRenditionIsHDR ? metadata.offsetSdr : metadata.offsetHdr;
+
+    float rgbMaxLinear = 0; // Max tone mapped pixel value across R, G and B channels.
+    float rgbSumLinear = 0; // Sum of max(r, g, b) for mapped pixels.
+
+    for (uint32_t j = 0; j < height; ++j) {
+        for (uint32_t i = 0; i < width; ++i) {
+            float basePixelRGBA[4];
+            avifGetRGBAPixel(baseImage, i, j, &baseRGBInfo, basePixelRGBA);
+            if (weight == 0.0 && outputTransferCharacteristics == transferCharacteristics) {
```

Comment:
Optional: Since this conditional is loop invariant, it would be good to check this outside the for loop and have a simpler loop for this case.

Reply:
Done, I moved it earlier to avoid e.g. rescaling the gain map if we don't even need it.